### ISSUE TITLE
Can't bind uniform buffer

### ISSUE THREAD

**Issue by tmielcza** [gfx-rs/gfx#1008]
I made a small gfx test that crashes with:
```
thread 'main' panicked at 'Error after executing command BindProgram(0): Invalid Enum'
```
when I try to create data for the draw pipeline:
```rust
    let data = pipe::Data {
        vertex_buffer: v_buf,
        some_uniform: [0., 0.],
        locals: factory.create_constant_buffer(1),
        out_color: target.clone(),
    };
```
After further investigation, I found the guilty call [here](https://github.com/gfx-rs/gfx/blob/d2af23073e303bbfc04ff10999db886dce88fc52/src/backend/gl/src/factory.rs#L126):
```rust
gl.BindBuffer(gl::BUFFER_UNIFORM, buffer);
```
I checked my OpenGL extensions and did not find ```GL_ARB_uniform_buffer_object```.
I'm on OpenGL 2.1.

**Comment by kvark** [comment]
Thanks! We should return an error during the pipeline creation when UBOs are not supported.

**Comment by tmielcza** [comment]
Ok. So, I can't use gfx on my laptop ? :'(

**Comment by kvark** [comment]
Just don't use constant buffers (`gfx::ConstantBuffer<>` component). Instead, pass stuff in globals (`gfx::Global<>` component).

**Comment by tmielcza** [comment]
Okay !
Also there's another bug, I don't know if it's related.
```thread 'main' panicked at 'gl function was not loaded'```
The function seems to be ```gl.BindFramebuffer()```.
```GL_ARB_framebuffer_object``` is missing.

**Comment by kvark** [comment]
Uh, sorry about it. This one is a blocker, I'll try to have a look tonight.

**Comment by tmielcza** [comment]
Thanks a lot!

**Comment by kvark** [comment]
Would you be able to get the latest version and run our examples to verify no more fixes are needed for your laptop? If you run the basics one just fine, I'll publish the fixed version of `gfx_device_gl`.

**Comment by tmielcza** [comment]
Works like a charm for almost every example.
The cube is black, though, I don't know why.
Thanks a lot for your quick intervention !

**Comment by tmielcza** [comment]
Here is the result of apitrace:  [trace.txt](https://github.com/gfx-rs/gfx/files/399177/trace.txt)
The log (I cut it down since It looped):  [log.txt](https://github.com/gfx-rs/gfx/files/399206/log.txt)
Incidentally, near all the examples run with a d3d backend. How can I run them using OpenGL?

**Comment by kvark** [comment]
DX11 is used by the example launcher by default on windows. If you want to force GL instead, try editing `main.rs` with the code similar to the one in the `shadow` example:
```rust
pub fn main() {
    <App<_, _> as gfx_app::ApplicationGL>::launch(
        "Multi-threaded shadow rendering example with gfx-rs",
        gfx_app::DEFAULT_CONFIG);
}
```
Also, if the black cube is Dx11-only issue, we need to open a new bug for it and investigate further. I don't have a Windows machine, and @sectopod 's one just died, so it might take some time, unless you want to take a stub at it yourself ;)

**Comment by msiglreith** [comment]
Concerning the `cube` example, manually recompiling the HLSL-shaders via `compile.cmd` fixed the issue on my system.

**Comment by kvark** [comment]
awesome! I recall someone doing the same thing too. If you can make a PR with those new shaders, that would be great.

**Comment by tmielcza** [comment]
I can confirm that works for me too.

**Comment by kvark** [comment]
The new fixes (except for the cube shaders) were published in `gfx_device_gl - 0.11.2`. You should need nothing more but a simple `cargo update` to proceed.
